File: 17-05-2022/Binary Tree Postorder Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<Integer> listt=new ArrayList<Integer>();
    public void postOrder(TreeNode root){
        if(root==null)
            return;
        postOrder(root.left);
        postOrder(root.right);
        listt.add(root.val);
    }
    public List<Integer> postorderTraversal(TreeNode root) {
        postOrder(root);
        return listt;
    }
}

File: 18-06-2022/Binary Tree Postorder Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        vector<int> res;
        stack<TreeNode*> s;
        TreeNode* curr=root;
        while(curr || !s.empty()){
            if(curr){
                s.push(curr);
                curr=curr->left;
            }
            else{
                if(s.empty())
                    break;
                TreeNode* node=s.top()->right;
                if(node){
                    curr=node;
                }
                else{
                    node=s.top();
                    s.pop();
                    res.push_back(node->val);
                    while(!s.empty() && node ==s.top()->right){
                        node=s.top();
                        s.pop();
                        res.push_back(node->val);
                    }
                }
            }
        }
        return res;
    }
};
